aten/src/ATen/Storage.h

```diff
@@ -19,29 +19,36 @@ struct AT_API Storage {
       size_t size,
       const std::function<void(void*)>& deleter,
       bool resizable = false);
-  ~Storage();
-  // There are reasonable interpretations of these constructors, but they're to
-  // be implemented on demand.
-  Storage(Storage&) = delete;
-  Storage(const Storage&) = delete;
-  Storage(Storage&&) = delete;
-  Storage(const Storage&&) = delete;
-  void set_pImpl(StorageImpl* storage_impl) {
-    storage_impl_ = storage_impl;
-  }
-  StorageImpl* pImpl() {
-    return storage_impl_;
+
+  template <typename T>
+  T* data() const { return storage_impl_->data<T>(); }
+
+  template <typename T>
+  T* unsafe_data() const { return storage_impl_->unsafe_data<T>(); }
+
+  size_t elementSize() const { return storage_impl_->elementSize(); }
+  ptrdiff_t size() const { return storage_impl_->size(); }
+  bool resizable() const { return storage_impl_->resizable(); }
+  // get() use here is to get const-correctness
+  void* data() const { return storage_impl_.get()->data(); }
+  const at::DataPtr& data_ptr() const { return storage_impl_->data_ptr(); }
+  DeviceType device_type() const { return storage_impl_->device_type(); }
+  at::Allocator* allocator() const { return storage_impl_.get()->allocator(); }
+  at::ScalarType scalar_type() const { return storage_impl_->scalar_type(); }
+  int getDevice() const { return storage_impl_->getDevice(); }
+
+  StorageImpl* unsafeReleaseStorageImpl() {
+    return storage_impl_.release();
```

Comment:
Why do we need this here over calling unsafeGetStorageImpl() followed by release at the call sites?

Reply:
Offline agreement: get rid of this method, and rewrite the one call site to manually incref itself.